### PR TITLE
Introduce support for WebSub Subscriber, Publisher and Hub

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
Introduce support for WebSub Subscriber, Publisher and Hub - https://www.w3.org/TR/websub/#title.

Ports initial implementation done [here](https://github.com/ballerina-lang/ballerina/compare/master...MaryamZi:websub-subscriber) to comply with the new model/syntax.

### WebSub Subscriber
Providing the capability to introduce a WebSub Subscriber service, allows easy introduction of webhooks with Ballerina.

Annotation based support is provided for developers to specify the hub they require subscribing at and the topic they require subscribing to along with other subscription parameters, based on which subscription requests would be sent once the service is started. Developers would now have in-built support to perform core WebSub functions related to subscription including intent verification and signature validation.

```ballerina
import ballerina.net.http;
import ballerina.net.websub;

endpoint websub:SubscriberServiceEndpoint websubEP {
    host:"localhost",
    port:8181
};

@websub:SubscriberServiceConfig {
    basePath:"/websub",
    subscribeOnStartUp:true,
    topic: "http://www.websubpubtopic.com",
    hub: "http://localhost:9999/websub/hub",
    leaseSeconds: 3600000,
    secret: "Kslk30SNF2AChs2"
}
service<websub:SubscriberService> websubSubscriber bind websubEP {

	onVerifyIntent (endpoint client, http:Request request) {
		http:Response response = {};
		response = websub:buildSubscriptionVerificationResponse(request);
		_ = client -> respond(response);
	}

	onNotification (endpoint client, http:Request request) {
		http:Response response = { statusCode:202 };
		_ = client -> respond(response);

                //process WebSub notification - including signature validation
		websub:WebSubNotification webSubNotification;
		websub:WebSubError webSubError;
		webSubNotification, webSubError = websub:processWebSubNotification(request);
		if (webSubError == null) {
			//signature validation successful
		} else {
			//error occurred processing notification
		}
	}

}
```

WebSub Subscriber Services could also be used with service providers who are not fully compliant or in partially WebSub compliant instances, to perform functions such as signature validation.

For example the following service would perform signature validation for JSON notifications from GitHub for a webhook registered via the UI, since GitHub send the X-Hub-Signature header.
```ballerina
import ballerina.net.http;
import ballerina.net.websub;

endpoint websub:SubscriberServiceEndpoint websubEP {
    host:"localhost",
    port:8181
};

@websub:SubscriberServiceConfig {
	basePath:"/githubwebhook",
	secret: "XKslk30SNF2AChs2"
}
service<websub:SubscriberService> githubwebhookListener bind websubEP {

	onNotification (endpoint client, http:Request request) {
		http:Response response = { statusCode:202 };
		_ = client -> respond(response);
		websub:WebSubNotification webSubNotification;
		websub:WebSubError webSubError;
		webSubNotification, webSubError = websub:processWebSubNotification(request);
		if (webSubError == null) {
			//signature validation successful
		} else {
			//error occurred processing notification
		}
	}

}
```
A client connector is also made available to send subscription/unsubscription requests to the hub explicitly.
```ballerina
websubHubClientEP -> subscribe(subscriptionChangeRequest);
```
```ballerina
websubHubClientEP -> unsubscribe(subscriptionChangeRequest);
```
For example, an unsubscription request could be sent as follows:
```ballerina
import ballerina.net.http;
import ballerina.net.websub;

endpoint websub:HubClientEndpoint websubHubClientEP {
    uri: "http://localhost:9999/websub/hub"
};

function main (string [] args) {
    websub:SubscriptionChangeRequest subscriptionChangeRequest = {topic:"http://www.websubpubtopic.com", callback:"http://localhost:8181/websub"};
    websub:SubscriptionChangeResponse subscriptionChangeResponse;
    websub:WebSubError webSubError;
    
    subscriptionChangeResponse, webSubError = websubHubClientEP -> unsubscribe(subscriptionChangeRequest);
    if (webSubError == null) {
        //unsubscription request successful 
    } else {
        //unsubscription request unsuccessful 
    }
}
```

### WebSub Hub
Ballerina comes with an in-built WebSub compliant Hub which developers can simply start up as required, providing Ballerina WebSub publishers a default Hub to use.
```ballerina
websub:WebSubHub webSubHub = websub:startUpHubService();
```
```ballerina
boolean stopped = webSubHub.shutdownBallerinaHub();
```

### WebSub Publisher
Ballerina WebSub Publishers can make use of functions made available to perform various WebSub related functions including publishing updates to the internal Ballerina WebSub Hub and building a response including link headers required for WebSub discovery (for services which support the same).
```ballerina
websub:WebSubHub webSubHub = websub:startUpBallerinaHub();
_ = webSubHub.publishUpdate("http://www.websubpubtopic.com", {"mode":"internal-hub"});
```
```ballerina
http:Response response = {};
response = websub:addWebSubLinkHeaders(response, ["http://localhost:9999/websub/hub", "http://localhost:9999/websub/hubtwo"], "http://www.websubpubtopic.com");
```

Publishers can also make use of the HubClientEndpoint to publish updates to remote Ballerina WebSub Hubs.

```ballerina
import ballerina.net.websub;

endpoint websub:HubClientEndpoint websubHubClientEP {
    uri: "http://localhost:9999/websub/hub"
};

function main (string [] args) {
    websub:WebSubError webSubError = websubHubClientEP -> publishUpdateToRemoteHub("http://www.websubpubtopic.com", {"mode":"remote-hub"});
    if (webSubError == null) {
        //unsubscription request successful 
    } else {
        //unsubscription request unsuccessful 
    }
}
```

## Approach
- Support for Websub subscriber services is introduced with the introduction of the WebSub subscriber service type.
- The WebSub Hub is introduced in Ballerina as a Ballerina service which uses the message broker to manage subscriptions and deliver content. The hub operates in in-memory mode by default (subscriptions are not persisted), and provides the option of enabling database based persistence of subscriptions if required.
- Support for Websub publishers is introduced via functions made available, which can be used by services or main programs.

## Automation tests
Added.

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
[Subscription and internal/remote hub publishing sample](https://github.com/MaryamZi/ballerina-websub-sample)

### CODE REVIEW

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/WebSubSubscriberConstants.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package org.ballerinalang.net.http;
+
+/**
+ * Constants for WebSubSubscriber Services.
+ *
+ * @since 0.965
```
Comment (chamil321) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/WebSubSubscriberConstants.java]: 0.965.0*

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/annotation.bal [hunk 2]
```diff
@@ -58,6 +58,26 @@ public struct WebSocketServiceConfiguration {
 @Field {value: "CONNECTION: Create a new instance of the service for each connection"}
 @Field {value: "SESSION: Create a new instance of the service for each session"}
 @Field {value: "SINGLETON: Create a single instance of the service and use it to process all requests coming to an endpoint"}
+
+@Description {value:"Configuration for a WebSubSubscriber service."}
+@Field {value: "endpoints: array of endpoints the service would be attached to"}
+@Field {value:"basePath: path of the WebSubSubscriber service"}
+@Field {value:"subscribeOnStartUp: whether a subscription request is expected to be sent on start up"}
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/annotation.bal]: Typos

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/websub_subscriber.bal [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        log:printError("Unable to verify intent since the topic is not specified");
+        return;
+    }
+
+    string topic = webSubSubscriberAnnotations.topic;
+    map params = request.getQueryParams();
+    var reqMode, _ = (string) params[HUB_MODE];
+    var challenge, _ = (string) params[HUB_CHALLENGE];
+    var reqTopic, _ = (string) params[HUB_TOPIC];
+    var reqLeaseSeconds, _ = (string) params[HUB_LEASE_SECONDS];
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/websub_subscriber.bal]: Why are the errors ignored here?

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/websub_subscriber.bal [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        generatedSignature = crypto:getHmac(stringPayload, secret, crypto:Algorithm.SHA256);
+    } else if (MD5.equalsIgnoreCase(method)) {
+        generatedSignature = crypto:getHmac(stringPayload, secret, crypto:Algorithm.MD5);
+    } else {
+        webSubError = {errorMessage:"Unsupported signature method: " + method};
+        return;
+    }
+
+    if (!signature.equalsIgnoreCase(generatedSignature)) {
+        webSubError = {errorMessage:"Signature validation failed: Invalid Signature!"};
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/websub_subscriber.bal]: Shall we include the signature as well?

**Pull request merged**